core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -642,6 +643,24 @@ private[log] class Cleaner(val id: Int,
   }
 
   /**
+   * Estimate what the last offset is for the current segment because we can't be sure
+   * that the index is returning a valid value in all cases.
+   *    @param segs - remaining segments to group. The current head is segment were estimating the lastOffset for.
+   *
+   * @return The estimated last offset in this segment
+   */
+  private def estimateLastOffset(segs: List[LogSegment], firstUncleanableOffset: Long): Long = {
+    if (segs.size > 1) {
+      /* if there is a next segment, use its base offset as the bounding offset to guarantee we know 
+       * the worst case offset */
+      segs(1).index.baseOffset - 1
+    } else {
+      //for the last segment in the list, use the first offset of the first uncleanable segment
```

Comment:
We can just say "for the last segment in the list, use the first uncleanable offset."